### ISSUE TITLE
Investigate compiler warnings in amdclang configuration

### ISSUE THREAD

**Issue by kennyweiss** [LLNL/axom#1024]
Our `amdclang` build on LLNL's EAS3 platform has several warnings, including:
* `-Wunused-parameter` and `-Wunused-variable` that seem to be related to our `AXOM_UNUSED_{PARAM,VAR}` macros. This might be related to problems related to our `AXOM_DEVICE_CODE` macro (see #734)
* Perhaps more significant are the `-Wgpu-maybe-wrong-side` warnings from some of our lambdas, e.g.
> <axom>/src/axom/core/../../axom/quest/IntersectionShaper.hpp:669:32: warning: capture host side class data member by this pointer in device or host device lambda function may result in invalid memory access if this pointer is not accessible on device side [-Wgpu-maybe-wrong-side]

Thanks @robinson96 for reporting the `-Wgpu-maybe-wrong-side` warnings

**Comment by BradWhitlock** [comment]
There are several instances of the same message during the compilation:
```
/usr/WS2/whitlocb/Axom/axom/src/axom/core/../../axom/quest/IntersectionShaper.hpp:834:17: warning: capture host side class data member by this pointer in device or host device lambda function may result in invalid memory access if this pointer is not accessible on device side [-Wgpu-maybe-wrong-side]
            if(!oct_has_duplicate_verts(local_octs[octIdx]))
```

**Comment by BradWhitlock** [comment]
The Lambda was capturing the "this" pointer of the IntersectionShaper to call the oct_has_duplicate_verts() method. It probably worked because the method was host/device and did not access anything through the this pointer. I changed the code so the has_duplicate_verts() method lives in Octahedron instead so the "this" pointer does not have to be captured.

That's the only instance of the -Wgpu-maybe-wrong-side warning that crops up in the build. there are a lot of unused parameter and other warnings though. I'm putting fixes in the branch: https://github.com/LLNL/axom/tree/bugfix/whitlock/1024_investigate_compiler_warnings

**Comment by BradWhitlock** [comment]
I fixed a few bits of code that caused error messages. To address most of the unused variable/parameter warnings, I ended up changing the SLIC macros in the HIP case. When AXOM_DEBUG was enabled and we build with HIP, it was inserting statements like ((void)(0)) to indicate a no-op and ignore any expression passed to the macros. Ignoring the expression is what caused the warnings though. I added some different no-op code that references the expression and any message while still more of less doing nothing.

**Comment by rhornung67** [comment]
@BradWhitlock this can be closed since you merged this https://github.com/LLNL/axom/pull/1079 Correct?

**Comment by BradWhitlock** [comment]
Yes, this PR closes #1079.

**Comment by kennyweiss** [comment]
I think the intersection test is failing on tioga as a result of #1079 (see #1089). 
@bmhan12 has a temporary workaround in #1095 that fixes the test, but brings back some amdclang warnings.

Should we reopen this issue as a reminder to fix these?
